disperser/batcher/txn_manager.go

```diff
@@ -0,0 +1,160 @@
+// same account.
+type TxnManager struct {
+	mu sync.Mutex
+
+	ethClient   common.EthClient
+	requestChan chan *TxnRequest
+	logger      common.Logger
+
+	queueSize          int
+	txnRefreshInterval time.Duration
+}
+
+func NewTxnManager(ethClient common.EthClient, queueSize int, txnRefreshInterval time.Duration, logger common.Logger) *TxnManager {
+	return &TxnManager{
+		ethClient:          ethClient,
+		requestChan:        make(chan *TxnRequest, queueSize),
+		logger:             logger,
+		queueSize:          queueSize,
+		txnRefreshInterval: txnRefreshInterval,
+	}
+}
+
+func (t *TxnManager) Start(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case req := <-t.requestChan:
+				receipt, err := t.monitorTransaction(ctx, req)
+				if err != nil {
+					req.HandleFailure(ctx, err)
+				} else {
+					req.HandleSuccess(ctx, receipt)
+				}
+			}
+		}
+	}()
+	t.logger.Info("started TxnManager")
+}
+
+// ProcessTransaction sends the transaction and queues the transaction for monitoring.
+// It returns an error if the transaction fails to be sent for reasons other than timeouts.
+// TxnManager monitors the transaction and resends it with a higher gas price if it is not mined without a timeout.
+func (t *TxnManager) ProcessTransaction(ctx context.Context, req *TxnRequest) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	gasTipCap, gasFeeCap, err := t.ethClient.GetLatestGasCaps(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get latest gas caps: %w", err)
+	}
+
+	txn, err := t.ethClient.UpdateGas(ctx, req.Tx, req.Value, gasTipCap, gasFeeCap)
+	if err != nil {
+		return fmt.Errorf("failed to update gas price: %w", err)
+	}
+	err = t.ethClient.SendTransaction(ctx, txn)
+	if err != nil {
+		return fmt.Errorf("failed to send txn (%s) %s: %w", req.Tag, req.Tx.Hash().Hex(), err)
+	}
+
+	t.requestChan <- req
+	return nil
+}
+
+// monitorTransaction monitors the transaction and resends it with a higher gas price if it is not mined without a timeout.
+// It returns an error if the transaction fails to be sent for reasons other than timeouts.
+func (t *TxnManager) monitorTransaction(ctx context.Context, req *TxnRequest) (*types.Receipt, error) {
+	for {
+		ctxWithTimeout, cancel := context.WithTimeout(ctx, t.txnRefreshInterval)
+		defer cancel()
+
+		receipt, err := t.ethClient.EnsureTransactionEvaled(
+			ctxWithTimeout,
+			req.Tx,
+			req.Tag,
+		)
+		if err == nil {
+			return receipt, nil
+		}
+
+		if errors.Is(err, context.DeadlineExceeded) {
+			t.logger.Warn("transaction not mined within timeout, resending with higher gas price", "tag", req.Tag, "txHash", req.Tx.Hash().Hex())
+			txn, err := t.speedUpTxn(ctxWithTimeout, req.Tx, req.Tag)
+			if err != nil {
+				t.logger.Error("failed to speed up transaction", "err", err)
+				return nil, err
+			}
+			err = t.ethClient.SendTransaction(ctx, txn)
```

Comment:
In this case, do the transactions that follow this one in the TxManager need to be updated with a new nonce? Will the pending transactions all fail?

Reply:
TxManager only monitors the oldest transaction for speed ups, and it uses existing nonce to replace the transaction.
Transactions that follow this one don't get updated and they keep their original nonces. 
Generally, TxManager doesn't manage nonces at all. Since nonce is specified in transaction object, it keeps being reused within TxManager.